### DISCUSSION TITLE
How to get `--stdin-source=@last_cmd_output jq | bat` to work ?

### DISCUSSION THREAD

**Discussion by akharrou** [kovidgoyal/kitty#4651] [Q&A] [answered]
**Describe the bug**

Hello, I'm trying to have the last command output be passed to `jq` followed by `bat`, i.e `jq | bat` (could also be `less`, or another pager). however I can't seem to get it right, all i get is the screen flicker, then nothing happens, even when i run it on valid JSON. any suggestions ?

**To Reproduce**

Here's all the things i've tried (one at a time):

```plaintext
# kitty.conf
map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output jq | bat
map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output 'jq | bat'
map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output sh -c 'jq | bat'
map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output jqbat
```

>- the last one is using a script called `jqbat` that I added to my path:
>
>    ```zsh
>    #!/bin/sh
>    
>    # if file
>    if [[ -f $1 ]]; then
>      cat "$1" | jq --color-output | bat
>    # if string
>    elif [[ ! -z $1 ]]; then
>      echo "$1" | jq --color-output | bat
>    # if stdin
>    else
>      jq --color-output | bat
>    fi
>    ```

then i do:

```zsh
$> echo '{ "name": "hello" }'
{ "name": "hello" }
```

then i trigger the mapping by pressing `shift+ctrl+j`.

**Environment details**
```
kitty 0.24.2 created by Kovid Goyal
Darwin Mac.local 21.1.0 Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64 x86_64
```

**Comment by kovidgoyal** [comment]
kitty -o "map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output sh -c 'jq | bat; read < /dev/tty'"

**Comment by page-down** [comment]
@kovidgoyal
This discussion still has a `bug` tag hanging on it.
These features of GitHub need to be better designed.

---

I found that `--hold` doesn't work.
Tried the latest commit and it is working, thanks.

```conf
map f1 launch --type=overlay --stdin-source=@last_cmd_output --hold sh -c "jq | bat -l json"
```

**Comment by akharrou** [comment, marked as answer]
Using `--hold`, shows stderr, and I found that it was just a matter of paths not being found when doing `sh -c 'jq | bat'`, because `sh` doesn't read certain initialization files, ones where I add certain paths.

works great with:

```zsh
map shift+ctrl+j launch --type=overlay --stdin-source=@last_cmd_output sh -c '/usr/local/bin/jq --color-output | /usr/local/bin/bat'
```

- adding `; read < /dev/tty'` made it hang after I `esc` the pager.
- `bat --language json` (`bat -l json`) is a good idea. there's also `jq --color-output`, if you prefer `jq`'s style

Thank you for the help

**Reply by akharrou** [reply]
Btw I love this feature of being able to pass shell history, or last command or the other ones, to programs, it's such **a great idea 👍🏼**
